src/Microsoft.Health.Fhir.Core/Features/Conformance/Providers/WellKnownConfigurationProvider.cs

```diff
@@ -0,0 +1,122 @@
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Core.Configs;
+using Microsoft.Health.Fhir.Core.Features.Conformance.Models;
+using Microsoft.Health.Fhir.Core.Features.Operations;
+using Microsoft.Health.Fhir.Core.Messages.Get;
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.Fhir.Core.Features.Conformance.Providers
+{
+    public class WellKnownConfigurationProvider : IWellKnownConfigurationProvider
+    {
+        private const string OpenIdConfigurationPath = ".well-known/openid-configuration";
+        private const string SmartConfigurationPath = ".well-known/smart-configuration";
+
+        private readonly SecurityConfiguration _securityConfiguration;
+        private readonly IHttpClientFactory _httpClientFactory;
+        private readonly ILogger<WellKnownConfigurationProvider> _logger;
+
+        public WellKnownConfigurationProvider(
+            IOptions<SecurityConfiguration> securityConfigurationOptions,
+            IHttpClientFactory httpClientFactory,
+            ILogger<WellKnownConfigurationProvider> logger)
+        {
+            _securityConfiguration = EnsureArg.IsNotNull(securityConfigurationOptions?.Value, nameof(securityConfigurationOptions));
+            _httpClientFactory = EnsureArg.IsNotNull(httpClientFactory, nameof(httpClientFactory));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        }
+
+        public bool IsSmartConfigured()
+        {
+            return !string.IsNullOrWhiteSpace(_securityConfiguration?.SmartAuthentication?.Authority);
+        }
+
+        public Task<OpenIdConfigurationResponse> GetOpenIdConfigurationAsync(CancellationToken cancellationToken)
+        {
+            return GetConfigurationAsync<OpenIdConfigurationResponse>(OpenIdConfigurationPath, cancellationToken);
+        }
+
+        public Task<GetSmartConfigurationResponse> GetSmartConfigurationAsync(CancellationToken cancellationToken)
+        {
+            return GetConfigurationAsync<GetSmartConfigurationResponse>(SmartConfigurationPath, cancellationToken);
+        }
+
+        private async Task<TConfiguration> GetConfigurationAsync<TConfiguration>(string configurationPath, CancellationToken cancellationToken)
+            where TConfiguration : class
+        {
+            Uri configurationUrl = GetConfigurationUrl(configurationPath);
+
+            if (configurationUrl != null)
+            {
+                _logger.LogInformation("Fetching configuration using well-known endpoint: {ConfigurationUrl}.", configurationUrl.AbsoluteUri);
+
+                using HttpClient client = _httpClientFactory.CreateClient();
+                using var smartConfigurationRequest = new HttpRequestMessage(HttpMethod.Get, configurationUrl);
+                HttpResponseMessage response = await client.SendAsync(smartConfigurationRequest, cancellationToken);
```

Comment:
If you leave Auth/Audience blank in appsettings, and simply run the OSS server, and then hit the .well-known/smart-configuration endpoint, you get an endless loop as the server makes a request against it's own endpoint :).

I also wonder if we can't rely on an Oauth compliant IDP and then provide the "smart" compliance ourselves?  I am concerned that there may be few full "smart" compliant IDPs and we may limit the usability of this if we can't augment Oauth to provide a well-known/smart-configuration endpoint from the FHIR service, which in turn relies on the .well-known/openid-configuration of the IDP.

Reply:
_If you leave Auth/Audience blank in appsettings, and simply run the OSS server, and then hit the .well-known/smart-configuration endpoint, you get an endless loop as the server makes a request against it's own endpoint :)._

I guess I'm not sure why this would happen. If you leave Authority blank, does the  `SecurityConfiguration.Authentication.Authority` value get populated some other way? The `GetConfigurationUrl` method below expects that if both values are null or empty, null should be returned here.

Reply:
If Security is true, but no Auth or Audience is specified we use an in memory Identity Provider which loads users, roles and permitted data actions from role.json.  This is for testing purposes.  It will publish an authorize and token endpoint.

Reply:
_I also wonder if we can't rely on an Oauth compliant IDP and then provide the "smart" compliance ourselves? I am concerned that there may be few full "smart" compliant IDPs and we may limit the usability of this if we can't augment Oauth to provide a well-known/smart-configuration endpoint from the FHIR service, which in turn relies on the .well-known/openid-configuration of the IDP._

I have a solution for this and we can talk about where this solution belongs (either this repository or a separate one). The main issue with SMART is that the response from .well-known/smart-configuration contains capabilities representing both the FHIR server and the IDP. The capabilities of the FHIR server and the IDP need to be combined by the implementor. This probably means that there are settings that need to be configured depending on the IDP.

Reply:
> If Security is true, but no Auth or Audience is specified we use an in memory Identity Provider which loads users, roles and permitted data actions from role.json. This is for testing purposes. It will publish an authorize and token endpoint.

I added a condition to address this specific case. Let me know what you think.